test/cpp/interop/xds_interop_client.cc

```diff
@@ -0,0 +1,242 @@
+std::set<XdsStatsWatcher*> watchers;
+// Mutex for global_request_id and watchers
+std::mutex mu;
+
+/** Records the remote peer distribution for a given range of RPCs. */
+class XdsStatsWatcher {
+ public:
+  XdsStatsWatcher(int start_id, int end_id)
+      : start_id_(start_id), end_id_(end_id), rpcs_needed_(end_id - start_id) {}
+
+  void RpcCompleted(int request_id, std::string peer) {
+    if (start_id_ <= request_id && request_id < end_id_) {
+      {
+        std::lock_guard<std::mutex> lk(m_);
+        if (peer.empty()) {
+          no_remote_peer_++;
+        } else {
+          rpcs_by_peer_[peer]++;
+        }
+        rpcs_needed_--;
+      }
+      cv_.notify_one();
+    }
+  }
+
+  void WaitForRpcStatsResponse(LoadBalancerStatsResponse* response,
+                               int timeout_sec) {
+    {
+      std::unique_lock<std::mutex> lk(m_);
+      cv_.wait_for(lk, std::chrono::seconds(timeout_sec),
+                   [this] { return rpcs_needed_ == 0; });
+      response->mutable_rpcs_by_peer()->insert(rpcs_by_peer_.begin(),
+                                               rpcs_by_peer_.end());
+      response->set_num_failures(no_remote_peer_ + rpcs_needed_);
+    }
+  }
+
+ private:
+  int start_id_;
+  int end_id_;
+  int rpcs_needed_;
+  std::map<std::string, int> rpcs_by_peer_;
+  int no_remote_peer_;
+  std::mutex m_;
+  std::condition_variable cv_;
+};
+
+class TestClient {
+ public:
+  TestClient(std::shared_ptr<Channel> channel)
+      : stub_(TestService::NewStub(channel)) {}
+
+  void UnaryCall() {
+    SimpleResponse response;
+    ClientContext context;
+
+    int saved_request_id;
+    {
+      std::lock_guard<std::mutex> lk(mu);
+      saved_request_id = ++global_request_id;
+    }
+    std::chrono::system_clock::time_point deadline =
+        std::chrono::system_clock::now() +
+        std::chrono::seconds(FLAGS_rpc_timeout_sec);
+    context.set_deadline(deadline);
+    Status status = stub_->UnaryCall(
+        &context, SimpleRequest::default_instance(), &response);
+
+    {
+      std::lock_guard<std::mutex> lk(mu);
+      for (auto watcher : watchers) {
+        watcher->RpcCompleted(saved_request_id, response.hostname());
+      }
+    }
+
+    if (FLAGS_print_response) {
+      if (status.ok()) {
+        std::cout << "Greeting: Hello world, this is " << response.hostname()
+                  << ", from " << context.peer() << std::endl;
+      } else {
+        std::cout << "RPC failed: " << status.error_code() << ": "
+                  << status.error_message() << std::endl;
+      }
+    }
+  }
+
+ private:
+  std::unique_ptr<TestService::Stub> stub_;
+};
+
+class LoadBalancerStatsServiceImpl : public LoadBalancerStatsService::Service {
+ public:
+  Status GetClientStats(ServerContext* context,
+                        const LoadBalancerStatsRequest* request,
+                        LoadBalancerStatsResponse* response) {
+    int start_id;
```

Comment:
more of a question, how is this server context triggered?  i don't see where LoadBalancerStatsRequest is built and GetClientStats is called

I understand the watcher will wait on the conditional var and then respond.  i am a bit confused how we even get into this server context

Reply:
This is the mechanism for the client to expose LB-related stats during the test. It's a gRPC server exposed by the client process and queried by an external test driver: This is implemented in https://github.com/grpc/grpc/pull/21869 (`run_xds_tests.py`) in OSS and a similar Python-based test driver in our google3 grpclb tests. The driver spins up the server(s) and a client process, manipulates the LB (traffic director) configuration as appropriate, and sends a gRPC request to the `GetClientStats` method to verify client-side LB decisions.